sdk/core/azure-core/azure/core/rest/_rest.py

```diff
@@ -0,0 +1,389 @@
+        return self._content
+
+    def __repr__(self):
+        # type: (...) -> str
+        content_type_str = (
+            ", Content-Type: {}".format(self.content_type) if self.content_type else ""
+        )
+        return "<HttpResponse: {} {}{}>".format(
+            self.status_code, self.reason, content_type_str
+        )
+
+class HttpResponse(_HttpResponseBase):  # pylint: disable=too-many-instance-attributes
+    """**Provisional** object that represents an HTTP response.
+
+    **This object is provisional**, meaning it may be changed in a future release.
+
+    It is returned from your client's `send_request` method if you pass in
+    an :class:`~azure.core.rest.HttpRequest`
+
+    >>> from azure.core.rest import HttpRequest
+    >>> request = HttpRequest('GET', 'http://www.example.com')
+    <HttpRequest [GET], url: 'http://www.example.com'>
+    >>> response = client.send_request(request)
+    <HttpResponse: 200 OK>
+
+    :keyword request: The request that resulted in this response.
+    :paramtype request: ~azure.core.rest.HttpRequest
+    :ivar int status_code: The status code of this response
+    :ivar mapping headers: The response headers
+    :ivar str reason: The reason phrase for this response
+    :ivar bytes content: The response content in bytes.
+    :ivar str url: The URL that resulted in this response
+    :ivar str encoding: The response encoding. Is settable, by default
+     is the response Content-Type header
+    :ivar str text: The response body as a string.
+    :ivar request: The request that resulted in this response.
+    :vartype request: ~azure.core.rest.HttpRequest
+    :ivar internal_response: The object returned from the HTTP library.
+    :ivar str content_type: The content type of the response
+    :ivar bool is_closed: Whether the network connection has been closed yet
+    :ivar bool is_stream_consumed: When getting a stream response, checks
+     whether the stream has been fully consumed
+    :ivar int num_bytes_downloaded: The number of bytes in your stream that
+     have been downloaded
+    """
+
+    def __enter__(self):
+        # type: (...) -> HttpResponse
+        return self
+
+    def close(self):
+        # type: (...) -> None
+        self.is_closed = True
+        self._internal_response.close()
+
+    def __exit__(self, *args):
+        # type: (...) -> None
+        self.close()
+
+    def read(self):
+        # type: (...) -> bytes
+        """
+        Read the response's bytes.
+
+        """
+        if self._content is None:
+            self._content = b"".join(self.iter_bytes())
+        return self.content
+
+    def iter_raw(self):
+        # type: () -> Iterator[bytes]
+        """Iterate over the raw response bytes
+        """
+        raise NotImplementedError()
+
+    def iter_bytes(self):
+        # type: () -> Iterator[bytes]
+        """Iterate over the response bytes
+        """
+        raise NotImplementedError()
+
+    def iter_text(self):
```

Comment:
Do you override iter_text for aiohttp?

If not, how can we make sure self.encoding works fine?

Reply:
I ended up moving parts of the encoding into the base class for encoding.
Basically, 1. if we have an application/json, following RFC 7159 and 7483, I default this to "utf-8". 2. if the response's content has been read in, I'll add an additional step of using chardet.detect to get the response's content. With this, the self.encoding should cover more corner cases with `iter_text`

Reply:
Is it possible iter_text is called w/o .read()? 

In this case, content is empty?

Reply:
yeah definitely. Basically for iter_bytes, iter_text, and iter_lines, if `content` exists, we'll yield the content to users. Otherwise, we'll call stream download generator and get the bytes from there.

This is the same behavior as httpx

Reply:
You calls iter_bytes or content?

It seems to me iter_bytes does not generate content?

Reply:
here's kind of the scenarios where you would use one vs. the other
1. I want the whole body on the response, and I want to access it all at once: Here you would call .content, similar to `.body()` on the old responses. You may need to call `.read()` first, to read in the response's bytes to content before
2. I want to iterate over the body: here, i call one of the `iter_` functions. This doesn't read the body into the response, but outputs it as an iterator. As more of an implementation detail: if the response's bytes are already read into `content`, we will iterate over the content in all of the `iter_` functions except `iter_raw` (consistent with httpx), and output them in chunks <= the block size we get from the transport

Reply:
I think we are on same page. So if the content is not read into the body, iter_text will not be able to detect the encoding from content, right?

Reply:
It still will be able. to do all of the checks, except the final fallback I added, which is the chardet.detect of the content